SI-4147 Add an implementation of `mutable.TreeMap`

ruippeixotog opened the pull request:
This commit contains an implementation of a mutable red-black tree with focus on performance. It also contains a new `mutable.TreeMap` Scala collection that is backed by the aforementioned tree. The common generic factories and traits related to mutable sorted maps didn't exist yet, so this commit also adds them.

Regarding performance, `TreeMap` overrides (from `MapLike` and `SortedMapLike`) all of the most common methods for maps and also those whose default implementations are asymptotically worse than direct red-black tree algorithms (e.g. `last`, `clear`).

The `rangeImpl` method of `TreeMap` returns an instance of `TreeMapView`, an inner class of `TreeMap`. This view is backed by the same `RedBlackTree.Tree` instance, and therefore changes to the original map are reflected in the view and vice-versa. The semantics of mutating a view by adding and removing keys outside the view's range are the same of the current `mutable.TreeSet`. A bit less focus was given on the performance of views - in particular, getting the `size` of a `TreeMapView` is O(n) on the number of elements inside the view bounds. That can be improved in the future.

In a future commit, `mutable.TreeSet` can be changed to be backed by this red-black tree implementation.

I didn't add unit tests for this collection because I think writing extensive JUnit tests for all the methods would be way too ineffective. Instead, I created a separate project (available [here](https://github.com/ruippeixotog/scala-mutable-treemap-tests)) containing a suite of ScalaCheck properties for both [`TreeMap`](https://github.com/ruippeixotog/scala-mutable-treemap-tests/blob/master/src/test/scala/TreeMapProperties.scala) and [`TreeMapView`](https://github.com/ruippeixotog/scala-mutable-treemap-tests/blob/master/src/test/scala/TreeMapViewProperties.scala) and a (rather unscientific, but somewhat conclusive) [performance comparison](https://github.com/ruippeixotog/scala-mutable-treemap-tests/blob/master/src/test/scala/TreeMapBenchmark.scala) between `immutable.TreeMap`, `mutable.TreeMap` and `java.util.TreeMap`. A sample result from this last test:

```
Total unique elements to insert/get/delete: 1000000

######
Insertion
######
--- scala.collection.immutable.TreeMap ---
Took 1653 ms
--- scala.collection.mutable.TreeMap ---
Took 1264 ms
--- java.util.TreeMap ---
Took 965 ms

######
Search
######
--- scala.collection.immutable.TreeMap ---
Took 961 ms
--- scala.collection.mutable.TreeMap ---
Took 932 ms
--- java.util.TreeMap ---
Took 899 ms

######
Deletion
######
--- scala.collection.immutable.TreeMap ---
Took 2622 ms
--- scala.collection.mutable.TreeMap ---
Took 1072 ms
--- java.util.TreeMap ---
Took 1044 ms
```

I understand if you want tests for this in the Scala repository nonetheless. If that's the case, please tell me what would be a sufficient test suite for this collection and I'll add it. Please also tell me if I forgot to do something, as it's my first "big" contribution to Scala :)

This should close [SI-4147](https://issues.scala-lang.org/browse/SI-4147).

ruippeixotog committed:
SI-4147 Add an implementation of `mutable.TreeMap`

This commit contains an implementation of a mutable red-black tree with focus on performance. It also contains a new `mutable.TreeMap` Scala collection that is backed by the aforementioned tree. The common generic factories and traits related to mutable sorted maps didn't exist yet, so this commit also adds them.

Regarding performance, `TreeMap` overrides (from `MapLike` and `SortedMapLike`) all of the most common methods for maps and also those whose default implementations are asymptotically worse than direct red-black tree algorithms (e.g. `last`, `clear`).

The `rangeImpl` method of `TreeMap` returns an instance of `TreeMapView`, an inner class of `TreeMap`. This view is backed by the same `RedBlackTree.Tree` instance, and therefore changes to the original map are reflected in the view and vice-versa. The semantics of mutating a view by adding and removing keys outside the view's range are the same of the current `mutable.TreeSet`. A bit less focus was given on the performance of views - in particular, getting the `size` of a `TreeMapView` is O(n) on the number of elements inside the view bounds. That can be improved in the future.

In a future commit, `mutable.TreeSet` can be changed to be backed by this red-black tree implementation.

retronym commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+ * An object containing the red-black tree implementation used by mutable `TreeMaps`. For performance reasons, this
+ * implementation uses `null` references to represent leaves instead of a sentinel node. Currently, the internal nodes
+ * do not store their subtree size - only the tree object keeps track of their size. Therefore, while obtaining the size
+ * of the whole tree is O(1), knowing the number of entries inside a range is O(n) on the size of the range.
```
Could you please move the implementation comment to `//` comment block, the doc comment should be tailored for the user of the API, not the maintainer. It would be good to keep some notes about performance in the API docs, though.

ruippeixotog replied:
Do you think it is worth keeping notes in the API docs about performance? Saying that it is a red-black tree tells users the complexity of most operations, and the note about size performance seems really only internal, as I do not provide a public method for getting the size of a range (`TreeMapView` uses the size of its iterator for calculating the collection size).

axel22 replied:
I'd say it's worthwhile to leave the complexity comments in the user API docs, and move everything else into a normal comment. Perhaps you could alternatively move the O(n) comment about the range size to the respective method.

retronym commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+ * @version 2.12
+ * @since 2.12
+ */
+object RedBlackTree {
```
I think this should be `private[collection]`, like `immutable.RedBlackTree`.

retronym commented on src/library/scala/collection/mutable/TreeMap.scala:
```diff
@@ -0,0 +1,163 @@
+ * @define Coll mutable.TreeMap
+ * @define coll mutable tree map
+ */
+class TreeMap[A, B] private (tree: RB.Tree[A, B])(implicit val ordering: Ordering[A])
```
I think this class should be final. We have deprecated inheritance from `immutable.TreeMap`.

ruippeixotog replied:
In this case I can't do that because `TreeMapView` extends `TreeMap`. This is needed because `rangeImpl` in `SortedMapLike` requires implementations to return the same class of the original (the `This` type argument).

When `rangeImpl` is called, the immutable version creates a new `immutable.TreeMap` with the same class of the original, and `mutable.TreeSet` has `from` and `until` bounds in the original class already. I didn't follow the example of `mutable.TreeSet` because that would require always checking the defined bounds on each call to tree map operations, regardless of whether it is a range view or the whole collection. As I think the most common use case for a tree map is to do operations on it without acting on views, I decided to separate the two "code bases".

ruippeixotog replied:
On a side note, aside from preventing possible performance optimizations, I think there is no problem with this class not being `final`. The `tree` field is private and therefore hidden from subclasses, so there is little room to breaking the internals. On the other hand, I can see some uses to extending this class, e.g. for augmenting it with domain-specific methods or decorate existing ones.

retronym replied:
You could mark it `sealed`, rather than `final`. I'll let @Ichoran comment on the specifics of why the immutable version is moving towards finality. I lean towards the "make concrete classes final, or design/document them for extension" school of thought, though admittedly the existing collections classes are more liberal than that.

retronym commented:
Could you please update the commit title to: "SI-4147 Add an implementation of `mutable.TreeMap`"

retronym commented:
We are in the process of setting up property-based testing for the collections to https://github.com/scala/scala-collections-laws. Perhaps @Ichoran will take your tests cases over to the repository.

However, Scalacheck tests are supported under `test/scalacheck`. I'd suggest to add your tests initially to there, and we can move later them if needed.

retronym commented:
Further review, please, by @Ichoran, and, if he has a few moments spare, @axel22.

retronym commented:
- [ ] Serialization. Add a @SerialVersionUID's to the classes that will be serialized and add a test for stability of serialization in `test/files/run/t8549.scala`.

retronym commented:
@ruippeixotog I forgot to say: welcome and thank you! This contribution looks to be of a really high standard already, looking forwards to having this data structure in our toolbox.

ruippeixotog commented:
Hi @retronym, thanks for the comments! I'll do the changes you mentioned above (please read my comments about making `TreeMap` final, though).

There is no folder `test/scalacheck` in the project, did you mean `test/files/scalacheck`? Can you tell me if there is a way to run only those tests using Ant or SBT?

retronym commented:
`./test/partest --scalacheck` or `./test/partest test/files/scalacheck/test.scala` will run them. `ant test.suite` includes them, too.

ruippeixotog commented:
@retronym, I have just updated the pull request with the changes you mentioned :)

However, I didn't know exactly what to do in `test/files/run/t8549.scala`. I suppose I should manually add entries there for `TreeMap`, but I don't know how to run those `scalac-hash` commands, as I don't have those binaries in my PATH. Can you guide me on what to do?

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+        cmp = ord.compare(key, x.key)
+        x = if (cmp < 0) x.left else x.right
+      }
+      if (cmp <= 0) y else successor(y)
```
If `node == Node(1, null, null)` and `key == 1`, then we have:

```
x = Node(1, null, null)
y = x
cmp = ord.compare(1, 1) // == 0
x = null
```

And we return `y == Node(1, null, null)`.
Is this correct? Shouldn't we return the first strictly greater node after `key`, i.e. `null` in this case.

axel22 replied:
If the `after` is inclusive, a comment for this method would help.

ruippeixotog replied:
Yes, `minAfter` and `maxBefore` follow the semantics of `from` and `until` in ranges. I agree that it may be misleading, I'll add a comment explaining that.

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+        cmp = ord.compare(key, x.key)
+        x = if (cmp < 0) x.left else x.right
+      }
+      if (cmp > 0) y else predecessor(y)
```
If we have `node == Node(1, null, null)` and `key == 1`, then we return `predecessor(y)`, which is `null` this time. It seems that `before` is not inclusive. Could we document that? It is not immediately obvious.

ruippeixotog replied:
Yes, I implemented those methods with from/until semantics because they are useful for range projections. I'll add a proper description for both methods.

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+
+  // ---- helpers ----
+
+  private[this] def successor[A, B](node: Node[A, B]): Node[A, B] = {
```
It would be helpful to document what a successor is, and give a few examples in the comments.

ruippeixotog replied:
I'll do that too. I think it suffices to say that it is the node that succeeds the argument `node` in an in-order traversal, do you think it also needs examples?

axel22 replied:
I think just saying that is fine!

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+    }
+  }
+
+  private[this] def predecessor[A, B](node: Node[A, B]): Node[A, B] = {
```
Same here.

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,502 @@
+  private[this] def fixAfterInsert[A, B](tree: Tree[A, B], node: Node[A, B]): Unit = {
+    var z = node
+    while (isRed(z.parent)) {
+      if (z.parent eq z.parent.parent.left) {
```
Is there a possibility of a `NullPointerException` here?
Looks like, if you call this from insert after adding a right child to the root node of the tree, i.e.:

```
root
  |
Node(1) // this is red, from the first insert
  |    \
null   Node(3)  <-- call fixAfterInsert here (this is z)
```

then `z.parent.parent == null`, no?

ruippeixotog replied:
Note that the root node is always black in a red-black tree. Therefore, if `isRed(z.parent)`, `z.parent` can't be the root node and therefore `z.parent.parent` can't be `null`.

axel22 replied:
Ah, ok - I now saw line 204, where you ensure this.

retronym commented:
`scalac-hash` is from https://github.com/retronym/libscala. It's just a shortcut to run old versions of Scala.

But in this case, just use `./build/quick/bin/scala[c]`. The instructions in that test aren't really applicable to new datastructures.

The intention of that test is to serialize the datastructure, record the results (in the base64 encoded strings in the test case itself), and then check that we can continue to deserialize that in the future.

ruippeixotog commented:
Roger! I just did the process and updated the pull request.

Ichoran commented:
This looks quite promising, but I'm at a conference and recovering from food poisoning so I'm not going to get to it for at least a couple more days.

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+  /**
+   * Transplant the node `from` to the place of node `to`. This is done by making `from` a child of `to`'s parent
+   * instead of `from` and making `from`'s parent the new parent of `to`. The children of `from` are left unchanged.
+   */
```
This comment should be:

```
"This is done by setting `from` as a child of `to`'s previous parent, and setting `from`'s parent to the `to`'s previous parent. The children of `from` are left unchanged."
```

In the code, you never set the `to.parent`.

axel22 commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+        if (y.parent eq z) xParent = y
+        else {
+          xParent = y.parent
+          transplant(tree, y, y.right)
```
Do you require a fix here - is it possible to get violate the no-2-subsequent red nodes invariant?

ruippeixotog replied:
At this point in the method, it is indeed possible that a red node becomes the child of another red node - it is intended behavior. In this transplant `y.right` takes the place of `y` and that situation can happen if both `y.parent` and `y.right` are red (note that breaking that invariant can only happen if `y` is originally black, information that is stored in `yIsRed`).

At the end of `delete`, `fixAfterDelete` is called precisely to fix that. It receives the node that took the place of `y` (which is set as `y.right` in line 237) and its new parent, and does the necessary rotations in order to reestablish the tree invariants, going up the tree possibly until the root node to do that.

axel22 commented on test/files/scalacheck/MutableTreeMap.scala:
```diff
@@ -0,0 +1,249 @@
+  implicit def arbTreeMap[A: Arbitrary: Ordering, B: Arbitrary] = Arbitrary(genTreeMap[A, B])
+}
+
+object MutableTreeMapProperties extends Properties("mutable.TreeMap") with MutableTreeMapGenerators {
```
Very nice set of checks!

However, what's missing here is checking the tree invariants. Could you add an assertion/property in each of the checks that verifies that each of the produced trees has the correct invariants?

(i.e. you defined `isValid` above, but you should use it in the properties)

ruippeixotog replied:
I could do that, but two problems arise:
- Right now, the `RedBlackTree` object is `private[collection]`, which means that I don't have access to that method in this file. I can easily declare this file to be in package `scala.collection.mutable` though, as some of the other scalacheck tests do, e.g. `avl.scala`;
- More importantly, I can't access the `RB.Tree` inside a `mutable.TreeMap` as it is private, and I think making it `private[collection]` just for this use case maybe isn't worth it.

For the second problem, what do you propose? I can implement another set of scalacheck properties where I test inserting and removing multiple elements on a `RB.Tree`, checking if it `isValid` on each step. I can also do mixed inserts/deletes on `mutable.TreeMap` and check if its iterator returns the correct entries on each step. I like the second approach more, as I tend to prefer blackbox testing on situations like this - trust that scalacheck finds the sequence of public operations that would break the internals, which leads to externally visible bugs (e.g. missing or duplicate entries). But I also see advantages on the whitebox approach and I can surely do tests directly on `RedBlackTree` if you want.

axel22 replied:
I would change package for the check file. I'd say package-private for RB.Tree is fine - we've done it in the past for other collection types.

Alternatively, you could duplicate `isValid` logic in the test.
What you propose about separate checks for invariants sounds reasonable to me.

ruippeixotog replied:
Hmm, ok! To check if I understood well, I'll create a separate `RedBlackTreeProperties` test class with property checks that adding and removing keys in any order will keep the invariants. I'll also add `++=` and `--=` property checks for the two existing `TreeMap` test classes. That way, I won't even need to make `tree` package-private. Is this ok? :)

axel22 replied:
Sounds good to me.

On May 29, 2015 12:27:33 AM GMT+02:00, "Rui Gonçalves" notifications@github.com wrote:

> > +trait MutableTreeMapGenerators {
> > +
> > -  // Note: in scalacheck 1.12.2 tree maps can be automatically
> >   generated without the need for custom
> > -  // machinery
> > -  def genTreeMap[A: Arbitrary: Ordering, B: Arbitrary]:
> >   Gen[mutable.TreeMap[A, B]] = {
> > -    import org.scalacheck.Gen._
> > -    for {
> > -      keys <- listOf(arbitrary[A])
> > -      values <- listOfN(keys.size, arbitrary[B])
> > -    } yield mutable.TreeMap(keys zip values: _*)
> > -  }
> >   +
> > -  implicit def arbTreeMap[A: Arbitrary: Ordering, B: Arbitrary] =
> >   Arbitrary(genTreeMap[A, B])
> >   +}
> >   +
> >   +object MutableTreeMapProperties extends
> >   Properties("mutable.TreeMap") with MutableTreeMapGenerators {
> 
> Hmm, ok! To check if I understood well, I'll create a separate
> `RedBlackTreeProperties` test class with property checks that adding
> and removing keys in any order will keep the invariants. I'll also add
> `++=` and `--=` property checks for the two existing `TreeMap` test
> classes. That way, I won't even need to make `tree` package-private. Is
> this ok? :)
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/4504/files#r31284301

## 

Sent from my Android device with K-9 Mail. Please excuse my brevity.

axel22 commented on src/library/scala/collection/mutable/SortedMap.scala:
```diff
@@ -0,0 +1,53 @@
+
+  override protected[this] def newBuilder: Builder[(A, B), SortedMap[A, B]] = SortedMap.newBuilder[A, B]
+
+  override def empty: SortedMap[A, B] = SortedMap.empty
```
Add newline.

axel22 commented on src/library/scala/collection/mutable/SortedMap.scala:
```diff
@@ -0,0 +1,53 @@
+  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): SortedMap[A, B1] =
+    clone().asInstanceOf[SortedMap[A, B1]] ++= xs.seq
+}
+
```
You might want to also instantiate an abstract class for the `SortedMap` trait - see example here:

https://github.com/scala/scala/blob/v2.11.5/src/library/scala/collection/mutable/Map.scala#L78

axel22 commented on test/files/scalacheck/MutableTreeMap.scala:
```diff
@@ -0,0 +1,249 @@
+    map -= k
+    !map.contains(k) && map.get(k) == None && map.size == newExpectedSize
+  }
+
```
I would add one more test with `++=` here, and a test with multiple `-=` calls, and then call `isValid` in each of them.
Sometimes complex combinations of several operations manage to invalidate an invariant.

axel22 commented:
.LGTM after comments are addressed.

ruippeixotog commented:
I have just updated the pull request with your latest comments. I ended up adding a separate `RedBlackTreeProperties` which checks invariants as keys are inserted and deleted in a `RB.Tree`. Please let me know if it looks good to you now :)

ruippeixotog commented:
I noticed the build failed. Is that related to my pull request? It doesn't look like that judging by the Jenkins output, but I may be interpreting it wrong...

retronym commented:
Please rebase on origin/2.12.x to get rid of the spurious errors. We've switched to buiding 2.12.x PRs with OpenJDK8, and we had to update a configuration file that filters out VM warnings from output before comparing against checkfiles.

Ichoran commented on src/library/scala/collection/mutable/SortedMap.scala:
```diff
@@ -0,0 +1,57 @@
+  override def +[B1 >: B](kv: (A, B1)): SortedMap[A, B1] = clone().asInstanceOf[SortedMap[A, B1]] += kv
+
+  override def +[B1 >: B](elem1: (A, B1), elem2: (A, B1), elems: (A, B1) *): SortedMap[A, B1] =
+    this + elem1 + elem2 ++ elems
```
This looks like it creates three full copies of the tree, one per `+` or `++`.  You'll need to use `clone` here too.

Ichoran commented on src/library/scala/collection/mutable/TreeMap.scala:
```diff
@@ -0,0 +1,166 @@
+  override def keysIterator = RB.keysIterator(tree)
+  override def valuesIterator = RB.valuesIterator(tree)
+
+  override def foreach[U](f: ((A, B)) => U) = RB.foreach(tree, f)
```
Better to have an explicit `: Unit` return-type annotation here for clarity.

Ichoran commented on src/library/scala/collection/mutable/TreeMap.scala:
```diff
@@ -0,0 +1,166 @@
+
+  override def foreach[U](f: ((A, B)) => U) = RB.foreach(tree, f)
+  override def transform(f: (A, B) => B) = { RB.transform(tree, f); this }
+  override def clear() = tree.root = null
```
`: Unit` would be clearer here also (esp. for people who still think in chained `=`).

Ichoran commented on src/library/scala/collection/mutable/TreeMap.scala:
```diff
@@ -0,0 +1,166 @@
+      }
+    }
+
+    override def foreach[U](f: ((A, B)) => U) = iterator.foreach(f)
```
Same deal with `: Unit`

Ichoran commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+   */
+  private[this] def blackHeight[A, B](node: Node[A, B]): Int = {
+    if (node eq null) 1
+    else if (node.red && (isRed(node.left) || isRed(node.right))) -1
```
This checks the "all red nodes have a black parent" property, not the height-consistency property.  Putting it in here is a little confusing, at least if you call the method "blackHeight".

Ichoran commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+  private[this] def blackHeight[A, B](node: Node[A, B]): Int = {
+    if (node eq null) 1
+    else if (node.red && (isRed(node.left) || isRed(node.right))) -1
+    else if ((node.left ne null) && (node.left.parent ne node)) -1
```
This line and the next check bidirectional linking is correct.  Also a different property than black-height.  (Not even one of the classic properties, just part of the definition of what it means to be a tree.)

Ichoran commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+      val lh = blackHeight(node.left)
+      if (lh == -1) -1
+      else {
+        val bh = blackHeight(node.right)
```
The logic here is really simple, but I think `bh` should be `rh` so it's symmetric?  Also, since shortcutting isn't really important (this isn't for performance!), wouldn't it be clearest to write it as

```
val lh = blackHeight(node.left)
val rh = blackHeight(node.right)
if (lh == -1 || lh != rh) -1
else if (isRed(node)) lh
else lh + 1
```

or something like that?

Ichoran commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,521 @@
+  // on the size of the range.
+
+  @SerialVersionUID(21575944040195605L)
+  final class Tree[A, B](var root: Node[A, B], var size: Int) extends Serializable
```
`Tree` really doesn't seem to be pulling its weight here.  All it does is hold an extra `var size` which could easily be held by the collection instead, and it enforces an obligatory extra pointer dereference and object in memory.

ruippeixotog replied:
Well, I started out by maintaining a `var node: Node[A, B]` in the collection itself, but it would cause  some annoyances in the API of the tree operations.

`insert` and `delete` may cause the root node to change and in the current implementation it is not always easy to keep and return a reference to the new root node (the root node can be changed inside `transplant`, for example). Moreover, those operations rely on querying and comparing nodes against the "current" root node, so I would have to have some way of passing a reference to the root node which was always up-to-date.

My solution was to create the explicit `Tree` object which despite bringing the extra pointer dereference keeps the code clean and the red-black tree implementation clearly separated from the `TreeMap` implementation (`mutable.TreeSet` could benefit from this implementation too, I believe).

Ichoran commented:
Looks pretty good so far!  Mostly style issues.  I'll run it through the collections laws shortly.

ruippeixotog commented:
@Ichoran, thank you for the review, I hope you are feeling better! I addressed all of your comments. I didn't pay much attention to the `isValid` code because it was mainly for me to check invariants during development, but now that it is used in the specs all your suggested fixes and improvements make sense.

Ichoran commented:
Hmmm, got a NPE in collections tests in `transformNodeNonNull`.  I'm trying to isolate it.

Ichoran commented:
`transform` gives an NPE when transforming  a single-element map because `transformNode` assumes that if `node` exists, `node.left` does also (when it passes it on to `transformNodeNonNull`).

Ichoran commented on src/library/scala/collection/mutable/RedBlackTree.scala:
```diff
@@ -0,0 +1,545 @@
+  def transform[A, B](tree: Tree[A, B], f: (A, B) => B): Unit = transformNode(tree.root, f)
+
+  private[this] def transformNode[A, B, U](node: Node[A, B], f: (A, B) => B): Unit =
+    if (node ne null) transformNodeNonNull(node.left, f)
```
Should this be `transformNodeNonNull(node, f)`?  The way it is gives a NPE and misses part of the tree.

Ichoran commented:
Once you fix the NPE (I commented on the line how to fix it), it passes all the collections-laws tests.  But note that this isn't very good at exhaustively testing addition and subtraction from the library.  That is done much better by test/junit/scala/collection/SetMapConsistencyTest.scala.  If you could add TreeMap there also, that would increase our confidence.  It does a fair bit of churning.  I haven't tested the view yet because the coverage of views is not very good anywhere.

Ichoran commented:
This should also close https://issues.scala-lang.org/browse/SI-6938 since that basically says "immutable RB trees use too much memory", and the most practical solution is to use a mutable variant.  (If someone decides this uses too much memory also, that'd be a different bug anyway.)

ruippeixotog commented:
I have just fixed the dumb error in `transform` and added the new map to `SetMapConsistencyTest`. The scalacheck specs also cover random insert/remove sequences, mostly on a more internal level and on mutating operations.

If you want, after this PR is accepted I can do a new one in which `mutable.TreeSet` starts using a `RB.Tree[A, Null]` as its underlying data structure - shouldn't [SI-6938](https://issues.scala-lang.org/browse/SI-6938) be closed only once that happened?

Ichoran commented:
You're right regarding SI-6938; it's a little more than just having a RB tree available.  One does have to use it!  If you would like to look into whether it's easy to switch the wrapping in a source-compatible way, that would be great.  If not, this is already a great contribution!

ruippeixotog commented:
At a first glance it should be easy to change the internals of `mutable.TreeSet` to use the mutable RB tree instead of the immutable one without breaking client code - the constructor that receives the tree, as well as all the internal methods and fields, are private. It most probably wouldn't be binary compatible though, and `mutable.TreeSet`s serialized using old Scala versions wouldn't be deserializable after the change. If that's acceptable for `2.12.x`, I can surely do that :)

I rebased this branch on `origin/2.12.x` as @retronym suggested, but it didn't seem to solve all build issues...

ruippeixotog commented:
/rebuild

adriaanm commented:
/rebuild

adriaanm commented:
Ping @Ichoran. The build issues turned out to be a hotspot bug (see https://github.com/scala/scala/pull/4578#issuecomment-115051647). Thanks for your patience! We'll do our best to get this merged for M2.

adriaanm commented:
Ping @axel22 -- were your comments addressed? (PS: please only _start_ your comment with LGTM to signal the PR is ready for merge unconditionally)

axel22 commented:
Yes, they were addressed.

axel22 commented:
LGTM

adriaanm commented:
Thanks!
